common/client/multi_node.go

```
@@ -40,165 +37,160 @@ var (
-		nodes:               nodes,
-		sendonlys:           sendonlys,
-		chainID:             chainID,
-		chainType:           chainType,
-		lggr:                logger.Sugared(lggr).Named("MultiNode").With("chainID", chainID.String()),
-		selectionMode:       selectionMode,
-		noNewHeadsThreshold: noNewHeadsThreshold,
-		nodeSelector:        nodeSelector,
-		chStop:              make(services.StopChan),
-		leaseDuration:       leaseDuration,
-		chainFamily:         chainFamily,
-		classifySendTxError: classifySendTxError,
-		reportInterval:      reportInterval,
-		sendTxSoftTimeout:   sendTxSoftTimeout,
+	c := &multiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT]{
+		primaryNodes:   primaryNodes,
+		sendOnlyNodes:  sendOnlyNodes,
+		chainID:        chainID,
+		lggr:           logger.Sugared(lggr).Named("MultiNode").With("chainID", chainID.String()),
+		selectionMode:  selectionMode,
+		nodeSelector:   nodeSelector,
+		chStop:         make(services.StopChan),
+		leaseDuration:  leaseDuration,
+		chainFamily:    chainFamily,
+		reportInterval: reportInterval,
 	}
 
 	c.lggr.Debugf("The MultiNode is configured to use NodeSelectionMode: %s", selectionMode)
 
 	return c
 }
 
+func (c *multiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT]) ChainType() config.ChainType {
+	return c.chainType
+}
+
+func (c *multiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT]) DoAll(ctx context.Context, do func(ctx context.Context, rpc RPC_CLIENT, isSendOnly bool) bool) error {
+	callsCompleted := 0
+	for _, n := range c.primaryNodes {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if n.State() != nodeStateAlive {
+			continue
+		}
+		if do(ctx, n.RPC(), false) {
+			callsCompleted++
+		}
+	}
+	for _, n := range c.sendOnlyNodes {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if n.State() != nodeStateAlive {
+			continue
+		}
+		if do(ctx, n.RPC(), false) {
+			callsCompleted++
```

There is no need to count sendOnyl calls. If we performed 0 calls to primary nodes, we should return an error